sklearn/metrics/pairwise.py

```diff
@@ -585,6 +592,119 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def gower_distances(X, Y=None, w=None, categorical_features=None):
+    """
+    Computes the gower distances between X and Y
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+
+    Y : array-like, shape (n_samples, n_features)
+
+    w:  array-like, shape (n_features)
+    According the Gower formula, w is an attribute weight.
+
+    categorical_features: array-like, shape (n_features)
+    Indicates with True/False wheter a column is a categorical attribute.
+    This is useful when categorical atributes are represented as integer
+    values.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, )
+
+    Notes
+    ------
+    Gower is a similarity measure for categorical, boolean and numerical mixed
+    data.
+
+    """
+
+    X, Y = check_pairwise_arrays(X, Y, dtype=(np.object, None)[issparse(X) or
+                                                               issparse(Y)])
+
+    rows, cols = X.shape
+
+    if categorical_features is None:
+        categorical_features = []
+        for col in range(cols):
+            if np.issubdtype(type(X[0, col]), np.number):
+                categorical_features.append(False)
+            else:
+                categorical_features.append(True)
+    # Calculates the normalized ranges and max values of numeric values
+    ranges_of_numeric = [0.0] * cols
+    max_of_numeric = [0.0] * cols
```

Comment:
why not just use `np.nanmax(X, axis=0)` in the dense array case?

And in the sparse case, you can replace NaN with -inf where necessary, and similarly calculate max.